File: ee-schkola/src-gen/main/kotlin/ee/schkola/person/PersonApiBase.kt
package ee.schkola.person

import ee.schkola.SchkolaBase
import java.util.*


enum class Gender {
    UNKNOWN,
    MALE,
    FEMALE;

    fun isUnknown() : Boolean = this == UNKNOWN
    fun isMale() : Boolean = this == MALE
    fun isFemale() : Boolean = this == FEMALE
}


fun String?.toGender(): Gender {
    return if (this != null) Gender.valueOf(this) else Gender.UNKNOWN
}


enum class GraduationLevel {
    UNKNOWN,
    MIDDLE_SCHOOL,
    SECONDARY_SCHOOL,
    HIGH_SCHOOL,
    TECHNICAL_COLLEGE,
    COLLEGE;

    fun isUnknown() : Boolean = this == UNKNOWN
    fun isMiddleSchool() : Boolean = this == MIDDLE_SCHOOL
    fun isSecondarySchool() : Boolean = this == SECONDARY_SCHOOL
    fun isHighSchool() : Boolean = this == HIGH_SCHOOL
    fun isTechnicalCollege() : Boolean = this == TECHNICAL_COLLEGE
    fun isCollege() : Boolean = this == COLLEGE
}


fun String?.toGraduationLevel(): GraduationLevel {
    return if (this != null) GraduationLevel.valueOf(this) else GraduationLevel.UNKNOWN
}


enum class MaritalState {
    UNKNOWN,
    SINGLE,
    MARRIED,
    SEPARATED,
    DIVORCED,
    WIDOWED;

    fun isUnknown() : Boolean = this == UNKNOWN
    fun isSingle() : Boolean = this == SINGLE
    fun isMarried() : Boolean = this == MARRIED
    fun isSeparated() : Boolean = this == SEPARATED
    fun isDivorced() : Boolean = this == DIVORCED
    fun isWidowed() : Boolean = this == WIDOWED
}


fun String?.toMaritalState(): MaritalState {
    return if (this != null) MaritalState.valueOf(this) else MaritalState.UNKNOWN
}




open class Address {
    val street: String
    val suite: String
    val city: String
    val code: String
    val country: String


    constructor(street: String = "", suite: String = "", city: String = "", code: String = "", country: String = "") {
        this.street = street
        this.suite = suite
        this.city = city
        this.code = code
        this.country = country
    }

    companion object {
        val EMPTY = Address()
    }
}


open class Church : SchkolaBase {
    val name: String
    val address: Address
    val pastor: PersonName
    val contact: Contact


    constructor(id: String = "", name: String = "", address: Address = Address(), pastor: PersonName = PersonName(), 
                contact: Contact = Contact()) : super(id) {
        this.name = name
        this.address = address
        this.pastor = pastor
        this.contact = contact
    }

    companion object {
        val EMPTY = Church()
    }
}


open class ChurchInfo {
    val church: String
    val association: String
    val member: Boolean
    val services: MutableList<String>


    constructor(church: String = "", association: String = "", member: Boolean = false, services: MutableList<String> = arrayListOf()) {
        this.church = church
        this.association = association
        this.member = member
        this.services = services
    }

    companion object {
        val EMPTY = ChurchInfo()
    }
}


open class Contact {
    val phone: String
    val email: String
    val cellphone: String


    constructor(phone: String = "", email: String = "", cellphone: String = "") {
        this.phone = phone
        this.email = email
        this.cellphone = cellphone
    }

    companion object {
        val EMPTY = Contact()
    }
}


open class Education {
    val graduation: Graduation
    val profession: String


    constructor(graduation: Graduation = Graduation(), profession: String = "") {
        this.graduation = graduation
        this.profession = profession
    }

    companion object {
        val EMPTY = Education()
    }
}


open class Family {
    val maritalState: MaritalState
    val childrenCount: Int
    val partner: PersonName


    constructor(maritalState: MaritalState = MaritalState.UNKNOWN, childrenCount: Int = 0, partner: PersonName = PersonName()) {
        this.maritalState = maritalState
        this.childrenCount = childrenCount
        this.partner = partner
    }

    companion object {
        val EMPTY = Family()
    }
}


open class Graduation : SchkolaBase {
    val name: String
    val level: GraduationLevel


    constructor(id: String = "", name: String = "", level: GraduationLevel = GraduationLevel.UNKNOWN) : super(id) {
        this.name = name
        this.level = level
    }

    companion object {
        val EMPTY = Graduation()
    }
}


open class PersonName {
    val first: String
    val last: String


    constructor(first: String = "", last: String = "") {
        this.first = first
        this.last = last
    }

    companion object {
        val EMPTY = PersonName()
    }
}


open class Profile : SchkolaBase {
    val gender: Gender
    val name: PersonName
    val birthName: String
    val birthday: Date
    val address: Address
    val contact: Contact
    val photoData: ByteArray
    val photo: String
    val family: Family
    val church: ChurchInfo
    val education: Education


    constructor(id: String = "", gender: Gender = Gender.UNKNOWN, name: PersonName = PersonName(), birthName: String = "", 
                birthday: Date = Date(), address: Address = Address(), contact: Contact = Contact(), 
                photoData: ByteArray = ByteArray(0), photo: String = "", family: Family = Family(), 
                church: ChurchInfo = ChurchInfo(), education: Education = Education()) : super(id) {
        this.gender = gender
        this.name = name
        this.birthName = birthName
        this.birthday = birthday
        this.address = address
        this.contact = contact
        this.photoData = photoData
        this.photo = photo
        this.family = family
        this.church = church
        this.education = education
    }

    companion object {
        val EMPTY = Profile()
    }
}

File: ee-schkola/src-gen/main/go/ee/schkola/person/PersonApiBase.go
package person

import (
    "ee/schkola"
    "github.com/eugeis/gee/enum"
    "time"
)

type Church struct {
    name string
    address *Address
    pastor *PersonName
    contact *Contact
    *schkola.SchkolaBase
}

func NewChurch(name string, address *Address, pastor *PersonName, contact *Contact, SchkolaBase *schkola.SchkolaBase) (ret *Church, err error) {
    ret = &Church{
        name: name,
        address: address,
        pastor: pastor,
        contact: contact,
        SchkolaBase: SchkolaBase,
    }
    
    return
}



type Graduation struct {
    name string
    level *GraduationLevel
    *schkola.SchkolaBase
}

func NewGraduation(name string, level *GraduationLevel, SchkolaBase *schkola.SchkolaBase) (ret *Graduation, err error) {
    ret = &Graduation{
        name: name,
        level: level,
        SchkolaBase: SchkolaBase,
    }
    
    return
}



type Profile struct {
    gender *Gender
    name *PersonName
    birthName string
    birthday *time.Time
    address *Address
    contact *Contact
    photoData []byte
    photo string
    family *Family
    church *ChurchInfo
    education *Education
    *schkola.SchkolaBase
}

func NewProfile(gender *Gender, name *PersonName, birthName string, birthday *time.Time, address *Address, contact *Contact, 
                photoData []byte, photo string, family *Family, church *ChurchInfo, education *Education, 
                SchkolaBase *schkola.SchkolaBase) (ret *Profile, err error) {
    ret = &Profile{
        gender: gender,
        name: name,
        birthName: birthName,
        birthday: birthday,
        address: address,
        contact: contact,
        photoData: photoData,
        photo: photo,
        family: family,
        church: church,
        education: education,
        SchkolaBase: SchkolaBase,
    }
    
    return
}







type Address struct {
    street string
    suite string
    city string
    code string
    country string
}

func NewAddress(street string, suite string, city string, code string, country string) (ret *Address, err error) {
    ret = &Address{
        street: street,
        suite: suite,
        city: city,
        code: code,
        country: country,
    }
    
    return
}



type ChurchInfo struct {
    church string
    association string
    member bool
    services string
}

func NewChurchInfo(church string, association string, member bool, services string) (ret *ChurchInfo, err error) {
    ret = &ChurchInfo{
        church: church,
        association: association,
        member: member,
        services: services,
    }
    
    return
}



type Contact struct {
    phone string
    email string
    cellphone string
}

func NewContact(phone string, email string, cellphone string) (ret *Contact, err error) {
    ret = &Contact{
        phone: phone,
        email: email,
        cellphone: cellphone,
    }
    
    return
}



type Education struct {
    graduation *Graduation
    profession string
}

func NewEducation(graduation *Graduation, profession string) (ret *Education, err error) {
    ret = &Education{
        graduation: graduation,
        profession: profession,
    }
    
    return
}



type Family struct {
    maritalState *MaritalState
    childrenCount int
    partner *PersonName
}

func NewFamily(maritalState *MaritalState, childrenCount int, partner *PersonName) (ret *Family, err error) {
    ret = &Family{
        maritalState: maritalState,
        childrenCount: childrenCount,
        partner: partner,
    }
    
    return
}



type PersonName struct {
    first string
    last string
}

func NewPersonName(first string, last string) (ret *PersonName, err error) {
    ret = &PersonName{
        first: first,
        last: last,
    }
    
    return
}




type Gender struct {
	name  string
	ordinal int
}

func (o *Gender) Name() string {
    return o.name
}

func (o *Gender) Ordinal() int {
    return o.ordinal
}

func (o *Gender) IsUnknown() bool {
    return o == _genders.Unknown()
}

func (o *Gender) IsMale() bool {
    return o == _genders.Male()
}

func (o *Gender) IsFemale() bool {
    return o == _genders.Female()
}

type genders struct {
	values []*Gender
    literals []enum.Literal
}

var _genders = &genders{values: []*Gender{
    {name: "Unknown", ordinal: 0},
    {name: "Male", ordinal: 1},
    {name: "Female", ordinal: 2}},
}

func Genders() *genders {
	return _genders
}

func (o *genders) Values() []*Gender {
	return o.values
}

func (o *genders) Literals() []enum.Literal {
	if o.literals == nil {
		o.literals = make([]enum.Literal, len(o.values))
		for i, item := range o.values {
			o.literals[i] = item
		}
	}
	return o.literals
}

func (o *genders) Unknown() *Gender {
    return _genders.values[0]
}

func (o *genders) Male() *Gender {
    return _genders.values[1]
}

func (o *genders) Female() *Gender {
    return _genders.values[2]
}

func (o *genders) ParseGender(name string) (ret *Gender, ok bool) {
	if item, ok := enum.Parse(name, o.literals); ok {
		return item.(*Gender), ok
	}
	return
}


type GraduationLevel struct {
	name  string
	ordinal int
}

func (o *GraduationLevel) Name() string {
    return o.name
}

func (o *GraduationLevel) Ordinal() int {
    return o.ordinal
}

func (o *GraduationLevel) IsUnknown() bool {
    return o == _graduationLevels.Unknown()
}

func (o *GraduationLevel) IsMiddleSchool() bool {
    return o == _graduationLevels.MiddleSchool()
}

func (o *GraduationLevel) IsSecondarySchool() bool {
    return o == _graduationLevels.SecondarySchool()
}

func (o *GraduationLevel) IsHighSchool() bool {
    return o == _graduationLevels.HighSchool()
}

func (o *GraduationLevel) IsTechnicalCollege() bool {
    return o == _graduationLevels.TechnicalCollege()
}

func (o *GraduationLevel) IsCollege() bool {
    return o == _graduationLevels.College()
}

type graduationLevels struct {
	values []*GraduationLevel
    literals []enum.Literal
}

var _graduationLevels = &graduationLevels{values: []*GraduationLevel{
    {name: "Unknown", ordinal: 0},
    {name: "MiddleSchool", ordinal: 1},
    {name: "SecondarySchool", ordinal: 2},
    {name: "HighSchool", ordinal: 3},
    {name: "TechnicalCollege", ordinal: 4},
    {name: "College", ordinal: 5}},
}

func GraduationLevels() *graduationLevels {
	return _graduationLevels
}

func (o *graduationLevels) Values() []*GraduationLevel {
	return o.values
}

func (o *graduationLevels) Literals() []enum.Literal {
	if o.literals == nil {
		o.literals = make([]enum.Literal, len(o.values))
		for i, item := range o.values {
			o.literals[i] = item
		}
	}
	return o.literals
}

func (o *graduationLevels) Unknown() *GraduationLevel {
    return _graduationLevels.values[0]
}

func (o *graduationLevels) MiddleSchool() *GraduationLevel {
    return _graduationLevels.values[1]
}

func (o *graduationLevels) SecondarySchool() *GraduationLevel {
    return _graduationLevels.values[2]
}

func (o *graduationLevels) HighSchool() *GraduationLevel {
    return _graduationLevels.values[3]
}

func (o *graduationLevels) TechnicalCollege() *GraduationLevel {
    return _graduationLevels.values[4]
}

func (o *graduationLevels) College() *GraduationLevel {
    return _graduationLevels.values[5]
}

func (o *graduationLevels) ParseGraduationLevel(name string) (ret *GraduationLevel, ok bool) {
	if item, ok := enum.Parse(name, o.literals); ok {
		return item.(*GraduationLevel), ok
	}
	return
}


type MaritalState struct {
	name  string
	ordinal int
}

func (o *MaritalState) Name() string {
    return o.name
}

func (o *MaritalState) Ordinal() int {
    return o.ordinal
}

func (o *MaritalState) IsUnknown() bool {
    return o == _maritalStates.Unknown()
}

func (o *MaritalState) IsSingle() bool {
    return o == _maritalStates.Single()
}

func (o *MaritalState) IsMarried() bool {
    return o == _maritalStates.Married()
}

func (o *MaritalState) IsSeparated() bool {
    return o == _maritalStates.Separated()
}

func (o *MaritalState) IsDivorced() bool {
    return o == _maritalStates.Divorced()
}

func (o *MaritalState) IsWidowed() bool {
    return o == _maritalStates.Widowed()
}

type maritalStates struct {
	values []*MaritalState
    literals []enum.Literal
}

var _maritalStates = &maritalStates{values: []*MaritalState{
    {name: "Unknown", ordinal: 0},
    {name: "Single", ordinal: 1},
    {name: "Married", ordinal: 2},
    {name: "Separated", ordinal: 3},
    {name: "Divorced", ordinal: 4},
    {name: "Widowed", ordinal: 5}},
}

func MaritalStates() *maritalStates {
	return _maritalStates
}

func (o *maritalStates) Values() []*MaritalState {
	return o.values
}

func (o *maritalStates) Literals() []enum.Literal {
	if o.literals == nil {
		o.literals = make([]enum.Literal, len(o.values))
		for i, item := range o.values {
			o.literals[i] = item
		}
	}
	return o.literals
}

func (o *maritalStates) Unknown() *MaritalState {
    return _maritalStates.values[0]
}

func (o *maritalStates) Single() *MaritalState {
    return _maritalStates.values[1]
}

func (o *maritalStates) Married() *MaritalState {
    return _maritalStates.values[2]
}

func (o *maritalStates) Separated() *MaritalState {
    return _maritalStates.values[3]
}

func (o *maritalStates) Divorced() *MaritalState {
    return _maritalStates.values[4]
}

func (o *maritalStates) Widowed() *MaritalState {
    return _maritalStates.values[5]
}

func (o *maritalStates) ParseMaritalState(name string) (ret *MaritalState, ok bool) {
	if item, ok := enum.Parse(name, o.literals); ok {
		return item.(*MaritalState), ok
	}
	return
}



